backend/app/notifications/model.py

```diff
@@ -1,36 +1,174 @@
+    )
+    data_output_dataset: Mapped["DataOutputDatasetAssociation"] = relationship(
+        "DataOutputDatasetAssociation",
+        back_populates="notifications",
+        passive_deletes=True,
+    )
+    __mapper_args__ = {
+        "polymorphic_identity": "DataOutputDatasetNotification",
+    }
+
+
+class DataProductMembershipNotification(Notification):
+    data_product_membership_id: Mapped[UUID] = mapped_column(
+        "data_product_membership_id",
+        ForeignKey("data_product_memberships.id", ondelete="CASCADE"),
+        nullable=False,
+    )
+    notification_origin: Mapped[DataProductMembershipStatus] = mapped_column(
+        Enum(DataProductMembershipStatus), nullable=False, use_existing_column=True
+    )
+    data_product_membership: Mapped["DataProductMembership"] = relationship(
+        "DataProductMembership",
+        back_populates="notifications",
+        passive_deletes=True,
+    )
+    __mapper_args__ = {
+        "polymorphic_identity": "DataProductMembershipNotification",
+    }
+
+
+class NotificationFactory:
+    @staticmethod
+    def createDataProductDatasetNotification(
+        db: Session, data_product_dataset: DataProductDatasetAssociation, approved: bool
+    ):
+        notification = DataProductDatasetNotification(
+            notification_type=NotificationTypes.DataProductDatasetNotification,
+            notification_origin=(
+                DataProductDatasetLinkStatus.APPROVED
+                if approved
+                else DataProductDatasetLinkStatus.DENIED
+            ),
+            data_product_dataset_id=data_product_dataset.id,
+        )
+        receivers = list(
+            {
+                owner.id: owner
+                for owner in (
+                    data_product_dataset.dataset.owners
+                    + [data_product_dataset.requested_by]
+                )
+            }.values()
+        )
+        notification.notification_interactions = [
+            NotificationInteraction(user=receiver, notification=notification)
+            for receiver in receivers
+        ]
+        db.add(notification)
+
+    @staticmethod
+    def createDataOutputDatasetNotification(
+        db: Session, data_output_dataset: DataOutputDatasetAssociation, approved: bool
+    ):
+        notification = DataOutputDatasetNotification(
+            notification_type=NotificationTypes.DataOutputDatasetNotification,
+            notification_origin=(
+                DataOutputDatasetLinkStatus.APPROVED
+                if approved
+                else DataOutputDatasetLinkStatus.DENIED
+            ),
+            data_output_dataset_id=data_output_dataset.id,
+        )
+        receivers = list(
+            {
+                owner.id: owner
+                for owner in (
+                    data_output_dataset.dataset.owners
+                    + [data_output_dataset.requested_by]
+                )
+            }.values()
+        )
```

Comment:
What I believe you are trying to achieve here is to create a collection of unique receivers? You map the user field in your `NotificationInteraction` object below, but I believe you can achieve the same by assigning the correct `user_id`. You can simplify your code to:

```suggestion
        import itertools
        
        receivers = set(
                owner.id for owner in (
                    itertools.chain(*data_output_dataset.dataset.owners, data_output_dataset.requested_by)
                )
            )
```

Reply:
Thank you for the suggestion, that is a lot cleaner indeed